client2/config/config.go

```diff
@@ -0,0 +1,341 @@
+	// we are willing to wait for the retreival of the PKI document.
+	InitialMaxPKIRetrievalDelay int
+
+	// PollingInterval is the interval in seconds that will be used to
+	// poll the receive queue.  By default this is 10 seconds.  Reducing
+	// the value too far WILL result in unnecessary Provider load, and
+	// increasing the value too far WILL adversely affect large message
+	// transmit performance.
+	PollingInterval int
+
+	// EnableTimeSync enables the use of skewed remote provider time
+	// instead of system time when available.
+	EnableTimeSync bool
+}
+
+func (d *Debug) fixup() {
+	if d.PollingInterval == 0 {
+		d.PollingInterval = defaultPollingInterval
+	}
+	if d.InitialMaxPKIRetrievalDelay == 0 {
+		d.InitialMaxPKIRetrievalDelay = defaultInitialMaxPKIRetrievalDelay
+	}
+	if d.SessionDialTimeout == 0 {
+		d.SessionDialTimeout = defaultSessionDialTimeout
+	}
+}
+
+// UpstreamProxy is the outgoing connection proxy configuration.
+type UpstreamProxy struct {
+	// Type is the proxy type (Eg: "none"," socks5").
+	Type string
+
+	// Network is the proxy address' network (`unix`, `tcp`).
+	Network string
+
+	// Address is the proxy's address.
+	Address string
+
+	// User is the optional proxy username.
+	User string
+
+	// Password is the optional proxy password.
+	Password string
+}
+
+func (uCfg *UpstreamProxy) toProxyConfig() (*proxy.Config, error) {
+	// This is kind of dumb, but this is the cleanest way I can think of
+	// doing this.
+	cfg := &proxy.Config{
+		Type:     uCfg.Type,
+		Network:  uCfg.Network,
+		Address:  uCfg.Address,
+		User:     uCfg.User,
+		Password: uCfg.Password,
+	}
+	if err := cfg.FixupAndValidate(); err != nil {
+		return nil, err
+	}
+	return cfg, nil
+}
+
+// Provider describes all necessary Provider connection information
+// so that clients can connect to the Provider and use the mixnet
+// and retrieve cached PKI documents.
+type Provider struct {
+	// Name is the human readable (descriptive) node identifier.
+	Name string
+
+	// IdentityKey is the node's identity (signing) key.
+	IdentityKey sign.PublicKey
+
+	// LinkKey is the node's wire protocol public key.
+	LinkKey wire.PublicKey
+
+	// Addresses is the map of transport to address combinations that can
+	// be used to reach the node.
+	Addresses map[string][]string
+}
+
+func (p *Provider) UnmarshalTOML(v interface{}) error {
+	_, p.IdentityKey = cert.Scheme.NewKeypair()
+	_, p.LinkKey = wire.DefaultScheme.GenerateKeypair(rand.Reader)
+
+	data, _ := v.(map[string]interface{})
+	p.Name = data["Name"].(string)
+	err := p.IdentityKey.UnmarshalText([]byte(data["IdentityKey"].(string)))
+	if err != nil {
+		return err
+	}
+	err = p.LinkKey.UnmarshalText([]byte(data["LinkKey"].(string)))
+	if err != nil {
+		return err
+	}
+
+	m := data["Addresses"].(map[string]interface{})
+	p.Addresses = make(map[string][]string)
+
+	for k, v := range m {
+		values := make([]string, 0)
+		if v == nil {
+			return fmt.Errorf("error: KEY %s has nil value\n", k)
```

Comment:
`k` here is inside `Addresses` inside the provider `p` ?
```suggestion
			return fmt.Errorf("error: KEY Providers.%s.[Addresses].%s has nil value\n", p.Name, k)
```